### DISCUSSION TITLE
Consider supporting semantic versioning

### DISCUSSION THREAD

**Discussion by Hnasar** [pyqtgraph/pyqtgraph#2245] [General]
### Short description

pyqtgraph releases bug fixes, features, and experimental changes in every release, which means users are unable to only subscribe to bug fixes updates

### Expected behavior

https://semver.org/ has one example overview. In general, breaking changes would go in major updates (e.g. 1.x.y -> 2.0.0); feature changes would go in minor updates (1.1.2 -> 1.2.0); bug fixes would go in patch updates (0.12.3 -> 0.12.4).

### Additional context

We are auditing our application dependencies, and would love to specify a requirement like:

```
pyqtgraph ~= 0.12
```

To get all non-breaking feature and bug fix updates, or like:

```
pyqtgraph ~= 0.12.3
```

To get all bug fix updates.

But right now both these are impossible..

Thank you for the nice library!

**Comment by j9ac9k** [comment]
Hi @Hnasar

There have been internal discussions the maintainers have had with respect to adopting semantic versioning.  I suppose I should start by saying, there is no consensus, but I will elaborate on some of the main reasons we have not adopted such a policy as of now.

1. We don't have remotely close to the number of maintainer/developer hours to be able to enforce this in a meaningful way.   The real wrinkle here is the "guarantee" that we won't break stuff for existing users with a patch release.  We can't even come close to that kind of assurance.
2. The distinction between bugfix (increment micro version) and enhancement (increment minor version) is a very hazy line, I promise you there will be vast disagreement on what changes qualify as what.  

@brettcannon (sorry for the ping, thought you should know you're being cited!) has a fantastic blog post about some of the issues with semantic versioning, which I must say, summarizes my feelings on the matter really nicely [here](https://snarky.ca/why-i-dont-like-semver/)

That said, our current version scheme is admittedly not great.  We are rolling the micro version when there is a significant change in versions of Qt bindings we support (removing Qt4 support, removing Qt5 < 5.12 support, etc), or some relatively large new feature.  We roll the patch version for more subtle bug fixes and/or performance improvements.  It's arbitrary and as you have pointed out, it makes doing appropriate pinning in applications much more difficult.

We are starting to lean towards using [Calendar Versioning](https://calver.org/).  A number of python projects have adopted this, and with chatting with maintainers, it has gone fairly well.  I don't know how calendar-based versioning would work with your project, but I would be curious to get your thoughts on it.

Also, if your application that uses pyqtgraph is available to the public, please consider submitting a PR updating the [Used By](https://github.com/pyqtgraph/pyqtgraph#used-by) section of the readme with the name and a link to it.

EDIT: this didn't feel right to have in the issue tracker, moving it to the discussions section.
